Create a temporary version of vmap that does not enforce output to be tensor only

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65566
* #65564
* __->__ #65565

This is to enable the next PR in the stack.

Alternative
 - Allow None to pass through for vmap. Discussing whether we want to do that https://github.com/facebookresearch/functorch/issues/159.
 

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

soulitzer committed:
Allow None to pass through for vmap

[ghstack-poisoned]

soulitzer committed:
Update on "Allow None to pass through for vmap"


Does jax allow this?

[ghstack-poisoned]

soulitzer committed:
Update on "Allow None to pass through for vmap"


Does jax allow this?

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236258).

soulitzer committed:
Update on "Allow None to pass through for vmap"


Does jax allow this?

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236258).

soulitzer committed:
Update on "Allow None to pass through for vmap"


This is to enable the next PR in the stack.

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

zou3519 commented on torch/_vmap_internals.py:
```diff
@@ -101,21 +101,21 @@ def _unwrap_batched(
         out_dim = out_dims_as_tuple[0]
         return torch._remove_batch_dim(batched_outputs, vmap_level, batch_size, out_dim)  # type: ignore[return-value]
-    return tuple(torch._remove_batch_dim(out, vmap_level, batch_size, out_dim)
+    return tuple((torch._remove_batch_dim(out, vmap_level, batch_size, out_dim) if out is not None else None)
```
Hmm, I don't know if this is the semantics we want for vmap. The rough contract for vmap is, in the absence of mutation, running `vmap(f)(x)` returns a stacked version of `[f(i) for i in x.unbind()]`.

Could you explain a bit more about the context behind this change?

zou3519 replied:
The interesting thing is that JAX's vmap allows None to pass through, so that's an argument for why we should do the same

soulitzer replied:
I messed up the PR ordering actually - this is meant to enable the previous PR, which was to allow vmap to directly wrap the call to the engine, which may sometimes return None when certain inputs are unused.

There doesn't seem to be an easy workaround since we'd either have to have a wrapper that:
 - either materializes the None into a dummy (zero?) tensor, then detect those and turn those back into `None` before returning
 - or has some side effect to keep track of where to insert the `None`s back.

Previously the application was to compute Jacobian in a vectorized way, and that was fine because we wanted to materialize the None into zeros anyway, but here in the more generic case we don't.

> Hmm, I don't know if this is the semantics we want for vmap. The rough contract for vmap is, in the absence of mutation, running vmap(f)(x) returns a stacked version of [f(i) for i in x.unbind()].

On the surface this contract seems a little restrictive. I feel like it would be convenient to allow users to return non-tensors and have those pass through. Is the reasoning behind this restriction to prevent user error?

zou3519 replied:
Interestingly, in JAX, non-tensors don't actually pass through (only None does). For example, if the function is
```
def f(x):
  return 0
  
vmap(f)(np.ones([10]))
```
then the result is a Tensor/array of 10 zeros.

zou3519 replied:
I'm still concerned about allowing None to pass through. In the vmap API [one passes an `out_dims` argument](https://github.com/facebookresearch/functorch/blob/1813f4a6d86ff9783996ac25b21c413511046793/functorch/_src/vmap.py#L172) that says "where the batch dim should appear in the output". The `None` object has no concept of dimensions though. Could you open an issue over at the [functorch repo](https://github.com/facebookresearch/functorch) please? I'm curious what other people think.

Here's a workaround for now: instead of modifying the vmap API, we can have our own "autograd-safe" vmap that allows Nones to pass through. It would probably look something like:

```py
def _vmap(func: Callable, in_dims: in_dims_t = 0, out_dims: out_dims_t = 0) -> Callable:
    @functools.wraps(func)
    def wrapped(*args, **kwargs):
        _check_out_dims_is_int_or_int_pytree(out_dims, func)
        batch_size, flat_in_dims, flat_args, args_spec = _process_batched_inputs(in_dims, args, func)
        vmap_level = _vmap_increment_nesting(batch_size)
        try:
            batched_inputs = _create_batched_inputs(flat_in_dims, flat_args, vmap_level, args_spec)
            batched_outputs = func(*batched_inputs, **kwargs)
            return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func, allow_none_passthrough=True)
        finally:
            _vmap_decrement_nesting()
    return wrapped
```

soulitzer committed:
Update on "Allow None to pass through for vmap"


This is to enable the next PR in the stack.

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

soulitzer committed:
Update on "Allow None to pass through for vmap"


This is to enable the next PR in the stack.

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

soulitzer committed:
Update on "Allow None to pass through for vmap"


This is to enable the next PR in the stack.

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

soulitzer committed:
Update on "Create a temporary version of vmap that does not enforce output to be tensor only"


This is to enable the next PR in the stack.

Alternative
 - Allow None to pass through for vmap. Discussing whether we want to do that https://github.com/facebookresearch/functorch/issues/159.
 

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

zou3519 commented on torch/_vmap_internals.py:
```diff
@@ -266,3 +270,21 @@ def wrapped(*args):
+# This is a version of vmap that allows None to pass through. It enables us to wrap the call
+# to the autograd engine, in `autograd.grad` which may return None if any of the inputs are
+# unused. See the issue discussing this: https://github.com/facebookresearch/functorch/issues/159.
+def _unsafe_vmap_for_autograd(func: Callable, in_dims: in_dims_t = 0, out_dims: out_dims_t = 0) -> Callable:
+    @functools.wraps(func)
+    def wrapped(*args):
+        _check_out_dims_is_int_or_int_tuple(out_dims, func)
+        vmap_level = torch._C._vmapmode_increment_nesting()
+        try:
+            batched_inputs, batch_size = _create_batched_inputs(in_dims, args, vmap_level, func)
+            batched_outputs = func(*batched_inputs)
+            # Don't validate the outputs
+            return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func, allow_none_pass_through=True)
+        finally:
+            torch._C._vmapmode_decrement_nesting()
```
Can we just combine this with the `_vmap` in this file by adding a `allow_none_pass_through` flag to it and documenting what it's used for? That one is also "private" (I don't remember what uses it)

soulitzer committed:
Update on "Create a temporary version of vmap that does not enforce output to be tensor only"


This is to enable the next PR in the stack.

Alternative
 - Allow None to pass through for vmap. Discussing whether we want to do that https://github.com/facebookresearch/functorch/issues/159.
 

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236258).

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236258).

soulitzer committed:
Update on "Create a temporary version of vmap that does not enforce output to be tensor only"


This is to enable the next PR in the stack.

Alternative
 - Allow None to pass through for vmap. Discussing whether we want to do that https://github.com/facebookresearch/functorch/issues/159.
 

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236258).

soulitzer committed:
Update on "Create a temporary version of vmap that does not enforce output to be tensor only"


This is to enable the next PR in the stack.

Alternative
 - Allow None to pass through for vmap. Discussing whether we want to do that https://github.com/facebookresearch/functorch/issues/159.
 

Differential Revision: [D31236258](https://our.internmc.facebook.com/intern/diff/D31236258)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236258).
